File: sdks/py/src/jeffs_brain_memory/memory/wikilink.py
# SPDX-License-Identifier: Apache-2.0
"""`[[topic]]` wikilink parsing and resolution."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING

from .paths import (
    memory_global_prefix,
    memory_project_prefix,
    project_slug as _project_slug,
)

if TYPE_CHECKING:
    from .manager import MemoryManager

WIKILINK_PATTERN = re.compile(r"\[\[([^\]]+)\]\]")


def extract_wikilinks(content: str) -> list[str]:
    return WIKILINK_PATTERN.findall(content)


def normalise_topic(topic: str) -> str:
    topic = topic.strip().lower().replace(" ", "-")
    return topic


def _resolve_in(mem: "MemoryManager", topic: str, prefix: str) -> str:
    normalised = normalise_topic(topic)
    if not normalised:
        return ""
    candidate = f"{prefix}/{normalised}.md"
    try:
        if mem.exists(candidate):
            return candidate
    except Exception:
        return ""
    return ""


def resolve_wikilink(mem: "MemoryManager", link: str, project_path: str) -> str:
    if "|" in link:
        link = link.split("|", 1)[0]
    link = link.strip()
    if not link:
        return ""

    slug = _project_slug(project_path)

    if link.startswith("global:"):
        topic = link[len("global:") :].strip()
        return _resolve_in(mem, topic, memory_global_prefix())

    resolved = _resolve_in(mem, link, memory_project_prefix(slug))
    if resolved:
        return resolved
    return _resolve_in(mem, link, memory_global_prefix())


def resolve_all_wikilinks(
    mem: "MemoryManager", content: str, project_path: str
) -> list[str]:
    links = extract_wikilinks(content)
    if not links:
        return []
    seen: set[str] = set()
    out: list[str] = []
    for link in links:
        resolved = resolve_wikilink(mem, link, project_path)
        if not resolved or resolved in seen:
            continue
        seen.add(resolved)
        out.append(resolved)
    return out

File: go/memory/wikilink.go
// SPDX-License-Identifier: Apache-2.0

package memory

import (
	"context"
	"regexp"
	"strings"

	"github.com/jeffs-brain/memory/go/brain"
)

// WikilinkPattern matches [[topic]], [[global:topic]], and
// [[topic|Display Text]] syntax.
var WikilinkPattern = regexp.MustCompile(`\[\[([^\]]+)\]\]`)

// ResolveWikilink resolves a wikilink reference to a logical brain path.
//
// For [[topic]], searches project memory first, falls back to global.
// For [[global:topic]], searches only global memory. The |Display Text
// part (if present) is stripped before resolution.
func (m *Memory) ResolveWikilink(ctx context.Context, link, projectPath string) brain.Path {
	if idx := strings.Index(link, "|"); idx >= 0 {
		link = link[:idx]
	}
	link = strings.TrimSpace(link)
	if link == "" {
		return ""
	}

	slug := ProjectSlug(projectPath)

	if strings.HasPrefix(link, "global:") {
		topic := strings.TrimSpace(strings.TrimPrefix(link, "global:"))
		return m.resolveTopicIn(ctx, topic, brain.MemoryGlobalPrefix())
	}

	if resolved := m.resolveTopicIn(ctx, link, brain.MemoryProjectPrefix(slug)); resolved != "" {
		return resolved
	}
	return m.resolveTopicIn(ctx, link, brain.MemoryGlobalPrefix())
}

// resolveTopicIn normalises a topic and checks whether the corresponding
// .md file exists under the given logical prefix.
func (m *Memory) resolveTopicIn(ctx context.Context, topic string, prefix brain.Path) brain.Path {
	normalised := normaliseTopic(topic)
	if normalised == "" {
		return ""
	}
	candidate := brain.Path(string(prefix) + "/" + normalised + ".md")
	exists, err := m.store.Exists(ctx, candidate)
	if err != nil || !exists {
		return ""
	}
	return candidate
}

// ResolveAllWikilinks resolves all wikilinks in content and returns the
// set of resolved paths (deduplicated).
func (m *Memory) ResolveAllWikilinks(ctx context.Context, content, projectPath string) []brain.Path {
	links := ExtractWikilinks(content)
	if len(links) == 0 {
		return nil
	}
	seen := make(map[brain.Path]struct{})
	var paths []brain.Path
	for _, link := range links {
		resolved := m.ResolveWikilink(ctx, link, projectPath)
		if resolved == "" {
			continue
		}
		if _, ok := seen[resolved]; ok {
			continue
		}
		seen[resolved] = struct{}{}
		paths = append(paths, resolved)
	}
	return paths
}

// ExtractWikilinks returns all wikilink references found in content.
func ExtractWikilinks(content string) []string {
	matches := WikilinkPattern.FindAllStringSubmatch(content, -1)
	if len(matches) == 0 {
		return nil
	}
	links := make([]string, 0, len(matches))
	for _, m := range matches {
		links = append(links, m[1])
	}
	return links
}

// normaliseTopic converts a human-readable topic name to a kebab-case
// filename stem.
func normaliseTopic(topic string) string {
	topic = strings.TrimSpace(topic)
	topic = strings.ToLower(topic)
	topic = strings.ReplaceAll(topic, " ", "-")
	return topic
}
